docs/DynamicCasting.md

```diff
@@ -0,0 +1,532 @@
+# Dynamic Casting Behavior
+
+* Author: [Tim Kientzle](https://github.com/tbkka)
+* Implementation: [apple/swift#29658](https://github.com/apple/swift/pull/29658)
+
+## Introduction
+
+The Swift language has three casting operators: `is`, `as?`, and `as!`.
+Each one takes an instance on the left-hand side and a type expression on the right-hand side.
+
+* The _cast test operator_ `is` tests whether a particular instance can be converted to a particular destination type.  It returns a boolean result.
+
+* The _conditional cast operator_ `as?` attempts the conversion and returns an `Optional` result: `nil` if the conversion failed, and a non-nil optional containing the result otherwise.
+
+* The _forced cast operator_ `as!` unconditionally performs the casting conversion and returns the result.  If an `as!` expression does not succeed, the implementation may terminate the program.
+
+Note: The static coercion operator `as` serves a different role and it's behavior will not be specifically discussed in this document.
+
+The following invariants relate the three casting operators:
+* Cast test: `x is T == ((x as? T) != nil)`
+* Conditional cast: `(x as? T) == (x is T) ? .some(x as! T) : nil`
+* Forced cast: `x as! T` is equivalent to `(x as? T)!`
+
+In particular, note that `is` and `as!` can be implemented in terms of `as?` and vice-versa.
+
+As with other operators with `!` in their names, `as!` is intended to only be invoked in cases where the programmer knows a priori that the cast will succeed.
+If the conversion would not succeed, then the behavior may not be fully deterministic.
+See the discussion of Array casting below for a specific case where this non-determinism can be important.
+
+The following sections detail the rules that govern the casting operations for particular Swift types.
+Except as described below, casting between dissimilar types always fails -- for example, casting a struct instance to a function type or vice-versa.
```

Comment:
Nit: "dissimilar types" could be clarified, since a formal "similarity" relation hasn't been defined. There's an intuitive idea of *structural incompatibility* between the different user-declarable type kinds, arrows, and tuples that is being captured here.

Reply:
Yes, I should have a simple list here of all the "kinds" of types.  (Is "kind" the right word?  Is there something better?)

Reply:
Actually, the sections themselves form such a list.  What if I changed this sentence to something along these lines:
```
Except where noted, casting between types described in different sections will always fail -- for example, ...
```

Reply:
Rephrased to try to be a bit more accurate (without diving down the rabbit hole of formal type theory).